<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial &mdash; mepylome 0.1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=a58bc63e"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API Reference" href="api.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            mepylome
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prerequisites">Prerequisites</a></li>
<li class="toctree-l2"><a class="reference internal" href="#setup">Setup</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parsing-idat-files">Parsing IDAT files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#manifest-files">Manifest files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rawdata">RawData</a></li>
<li class="toctree-l2"><a class="reference internal" href="#methyldata">MethylData</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#using-alternative-preprocessing-methods">Using Alternative Preprocessing Methods</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#copy-number-variation-cnv">Copy number variation (CNV)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">mepylome</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial">
<h1><a class="toc-backref" href="#id1" role="doc-backlink">Tutorial</a><a class="headerlink" href="#tutorial" title="Link to this heading"></a></h1>
<p>This tutorial will guide you through the steps to analyze methylation data
using the mepylome package. We’ll cover setting up the environment, parsing
IDAT files, working with manifest files, processing raw data, and performing
copy number variation (CNV) analysis. You can find the code of this tutorial
as a script in <code class="docutils literal notranslate"><span class="pre">mepylome/tests/tutorial_basic.py</span></code>.</p>
<section id="prerequisites">
<h2><a class="toc-backref" href="#id2" role="doc-backlink">Prerequisites</a><a class="headerlink" href="#prerequisites" title="Link to this heading"></a></h2>
<p>Before starting, ensure you have the mepylome package installed. You can
install it using pip:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>mepylome
</pre></div>
</div>
<nav class="contents" id="contents-of-tutorial">
<p class="topic-title">Contents of Tutorial</p>
<ul class="simple">
<li><p><a class="reference internal" href="#tutorial" id="id1">Tutorial</a></p>
<ul>
<li><p><a class="reference internal" href="#prerequisites" id="id2">Prerequisites</a></p></li>
<li><p><a class="reference internal" href="#setup" id="id3">Setup</a></p></li>
<li><p><a class="reference internal" href="#parsing-idat-files" id="id4">Parsing IDAT files</a></p></li>
<li><p><a class="reference internal" href="#manifest-files" id="id5">Manifest files</a></p></li>
<li><p><a class="reference internal" href="#rawdata" id="id6">RawData</a></p></li>
<li><p><a class="reference internal" href="#methyldata" id="id7">MethylData</a></p>
<ul>
<li><p><a class="reference internal" href="#using-alternative-preprocessing-methods" id="id8">Using Alternative Preprocessing Methods</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#copy-number-variation-cnv" id="id9">Copy number variation (CNV)</a></p></li>
</ul>
</li>
</ul>
</nav>
</section>
<section id="setup">
<h2><a class="toc-backref" href="#id3" role="doc-backlink">Setup</a><a class="headerlink" href="#setup" title="Link to this heading"></a></h2>
<p>First, import the required modules and set up the necessary directories:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>

<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mepylome</span> <span class="kn">import</span> <span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">CNV</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">IdatParser</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">Manifest</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">MethylData</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">RawData</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">ReferenceMethylData</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
</pre></div>
</div>
<p>Setup directories for the tutorial</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">DIR</span> <span class="o">=</span> <span class="n">Path</span><span class="o">.</span><span class="n">home</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;Documents&quot;</span> <span class="o">/</span> <span class="s2">&quot;mepylome&quot;</span> <span class="o">/</span> <span class="s2">&quot;projects&quot;</span> <span class="o">/</span> <span class="s2">&quot;basic&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ANALYSIS_DIR</span> <span class="o">=</span> <span class="n">DIR</span> <span class="o">/</span> <span class="s2">&quot;analysis&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">REFERENCE_DIR</span> <span class="o">=</span> <span class="n">DIR</span> <span class="o">/</span> <span class="s2">&quot;reference&quot;</span>
</pre></div>
</div>
<p>Download and setup the necessary data files (approximately 653 MB)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mepylome.tutorial</span> <span class="kn">import</span> <span class="n">setup_tutorial_files</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">setup_tutorial_files</span><span class="p">(</span><span class="n">DIR</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="parsing-idat-files">
<h2><a class="toc-backref" href="#id4" role="doc-backlink">Parsing IDAT files</a><a class="headerlink" href="#parsing-idat-files" title="Link to this heading"></a></h2>
<p>Define the IDAT file path</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idat_file</span> <span class="o">=</span> <span class="n">ANALYSIS_DIR</span> <span class="o">/</span> <span class="s2">&quot;200925700125_R07C01_Grn.idat&quot;</span>
</pre></div>
</div>
<p>Parse the IDAT file</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idat_data</span> <span class="o">=</span> <span class="n">IdatParser</span><span class="p">(</span><span class="n">idat_file</span><span class="p">)</span>
</pre></div>
</div>
<p>If you installed mepylome with C++ support (see installation) you can
also use the C++ parser (input must be a string, not a Path object)</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">mepylome</span> <span class="kn">import</span> <span class="n">_IdatParser</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">_idat_data</span> <span class="o">=</span> <span class="n">_IdatParser</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">idat_file</span><span class="p">))</span>
</pre></div>
</div>
<p>Display parsed data</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">idat_data</span><span class="p">)</span>
</pre></div>
</div>
<p>You should see the following output:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idat_data</span>
<span class="go">IdatParser(</span>
<span class="go">    file_size: 13686991</span>
<span class="go">    num_fields: 19</span>
<span class="go">    illumina_ids: array([ 1600101,  1600111, ..., 99810990, 99810992], dtype=int32)</span>
<span class="go">    probe_means: array([15629,  8469, ...,  7971,   943], dtype=uint16)</span>
<span class="go">    std_dev: array([1377,  408, ...,  702,  312], dtype=uint16)</span>
<span class="go">    n_beads: array([16,  7, ...,  6, 10], dtype=uint8)</span>
<span class="go">    mid_block: array([ 1600101,  1600111, ..., 99810990, 99810992], dtype=int32)</span>
<span class="go">    red_green: 0</span>
<span class="go">    mostly_null:</span>
<span class="go">    barcode: 200925700125</span>
<span class="go">    chip_type: BeadChip 8x5</span>
<span class="go">    mostly_a: R07C01</span>
<span class="go">    unknown_1:</span>
<span class="go">    unknown_2:</span>
<span class="go">    unknown_3:</span>
<span class="go">    unknown_4:</span>
<span class="go">    unknown_5:</span>
<span class="go">    unknown_6:</span>
<span class="go">    unknown_7:</span>
<span class="go">)</span>
</pre></div>
</div>
<p>The parsed data is available as attributes of the <code class="docutils literal notranslate"><span class="pre">IdatParser</span></code> object.</p>
<p>Access the Illumina IDs (probes IDs) from the parsed IDAT file</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ids</span> <span class="o">=</span> <span class="n">idat_data</span><span class="o">.</span><span class="n">illumina_ids</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="manifest-files">
<h2><a class="toc-backref" href="#id5" role="doc-backlink">Manifest files</a><a class="headerlink" href="#manifest-files" title="Link to this heading"></a></h2>
<p>Load the available manifest files for different array types.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">manifest_450k</span> <span class="o">=</span> <span class="n">Manifest</span><span class="p">(</span><span class="s2">&quot;450k&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">manifest_epic</span> <span class="o">=</span> <span class="n">Manifest</span><span class="p">(</span><span class="s2">&quot;epic&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">manifest_epic_v2</span> <span class="o">=</span> <span class="n">Manifest</span><span class="p">(</span><span class="s2">&quot;epicv2&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The first time you run this, the manifest files will be downloaded and
saved locally to ~/.mepylome. This initial download might take some time.</p>
</div>
<p>Obtain values from attributes</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">probes_df</span> <span class="o">=</span> <span class="n">manifest_450k</span><span class="o">.</span><span class="n">data_frame</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">controls_df</span> <span class="o">=</span> <span class="n">manifest_450k</span><span class="o">.</span><span class="n">control_data_frame</span>
</pre></div>
</div>
<p>Print overview</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">manifest_450k</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="rawdata">
<h2><a class="toc-backref" href="#id6" role="doc-backlink">RawData</a><a class="headerlink" href="#rawdata" title="Link to this heading"></a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">RawData</span></code> class handles both green and red signal intensity data from
IDAT files. You can initialize it using a base path to the IDAT files
(without the _Grn.idat / _Red.idat suffix), or by providing the full path to
either the Grn or Red IDAT file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idat_file</span> <span class="o">=</span> <span class="n">ANALYSIS_DIR</span> <span class="o">/</span> <span class="s2">&quot;200925700125_R07C01&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">raw_data</span> <span class="o">=</span> <span class="n">RawData</span><span class="p">(</span><span class="n">idat_file</span><span class="p">)</span>
</pre></div>
</div>
<dl class="simple">
<dt>The data is saved within the following attributes:</dt><dd><ul class="simple">
<li><p>raw_data.grn: Green signal intensities</p></li>
<li><p>raw_data.red: Red signal intensities</p></li>
<li><p>raw_data.array_type: Type of the array (e.g., 450k, EPIC)</p></li>
<li><p>raw_data.manifest: Corresponding manifest file</p></li>
<li><p>raw_data.ids: IDs on the bead</p></li>
</ul>
</dd>
</dl>
<p>Print an overview of the raw data</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">raw_data</span><span class="p">)</span>
</pre></div>
</div>
<p>This function also accepts basepaths (without the _Grn.idat / _Red.idat
ending:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idat_file</span> <span class="o">=</span> <span class="n">ANALYSIS_DIR</span> <span class="o">/</span> <span class="s2">&quot;200925700125_R07C01&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">raw_data</span> <span class="o">=</span> <span class="n">RawData</span><span class="p">(</span><span class="n">idat_file</span><span class="p">)</span>
</pre></div>
</div>
<p>RawData can also read multiple files, such as reference files:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">idat_file0</span> <span class="o">=</span> <span class="n">ANALYSIS_DIR</span> <span class="o">/</span> <span class="s2">&quot;200925700125_R07C01&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">idat_file1</span> <span class="o">=</span> <span class="n">ANALYSIS_DIR</span> <span class="o">/</span> <span class="s2">&quot;200925700133_R02C01&quot;</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">raw_data_2</span> <span class="o">=</span> <span class="n">RawData</span><span class="p">([</span><span class="n">idat_file0</span><span class="p">,</span> <span class="n">idat_file1</span><span class="p">])</span>
</pre></div>
</div>
<p>Alternatively, read all IDAT files in a directory (supports recursive
search):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">raw_data_all</span> <span class="o">=</span> <span class="n">RawData</span><span class="p">(</span><span class="n">ANALYSIS_DIR</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">raw_data</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="methyldata">
<h2><a class="toc-backref" href="#id7" role="doc-backlink">MethylData</a><a class="headerlink" href="#methyldata" title="Link to this heading"></a></h2>
<p>The raw data can be preprocessed using one of the following methods:
‘illumina’ (default), ‘swan’, or ‘noob’. Initialize MethylData with raw data
using the default ‘illumina’ preprocessing method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data_all</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data_all</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternatively, you can explicitly specify the ‘illumina’ preprocessing
method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="s2">&quot;illumina&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also initialize MethylData directly from an IDAT file path, without
using <code class="docutils literal notranslate"><span class="pre">RawData</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">idat_file</span><span class="p">)</span>
</pre></div>
</div>
<p>Obtain various values via the attributes of the MethylData object:</p>
<p>Access the beta values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">beta</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">beta</span>
</pre></div>
</div>
<p>Access the beta values at specific CpG sites (here at all the sites of the
EPICv2 manifest. You can fill missing values with a specified value (e.g.,
0.5).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">epicv2_cpgs</span> <span class="o">=</span> <span class="n">manifest_epic_v2</span><span class="o">.</span><span class="n">methylation_probes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">beta_specific</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">betas_for_cpgs</span><span class="p">(</span><span class="n">epicv2_cpgs</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
<p>Access the methylation signals for the green and red channels.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methylated_signals</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">methylated</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">unmethylated_signals</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">unmethylated</span>
</pre></div>
</div>
<p>Access the corrected color signals.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">corrected_green_signals</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">grn</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">corrected_red_signals</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">red</span>
</pre></div>
</div>
<p>Access the type of the array used (e.g., 450k, EPIC).</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">array_type</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">array_type</span>
</pre></div>
</div>
<p>Access the corresponding manifest file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">corresponding_manifest</span> <span class="o">=</span> <span class="n">methyl_data</span><span class="o">.</span><span class="n">manifest</span>
</pre></div>
</div>
<p>Print an overview of the methylation data.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">methyl_data</span><span class="p">)</span>
</pre></div>
</div>
<p>Preprocess the raw data using the SWAN or NOOB method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data_swan</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="s2">&quot;swan&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data_noob</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="s2">&quot;noob&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Print overview</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">methyl_data</span><span class="p">)</span>
</pre></div>
</div>
<section id="using-alternative-preprocessing-methods">
<h3><a class="toc-backref" href="#id8" role="doc-backlink">Using Alternative Preprocessing Methods</a><a class="headerlink" href="#using-alternative-preprocessing-methods" title="Link to this heading"></a></h3>
<p>Preprocess the raw data using the SWAN method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data_swan</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="s2">&quot;swan&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Preprocess the raw data using the NOOB method.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">methyl_data_noob</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">raw_data</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="s2">&quot;noob&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="copy-number-variation-cnv">
<h2><a class="toc-backref" href="#id9" role="doc-backlink">Copy number variation (CNV)</a><a class="headerlink" href="#copy-number-variation-cnv" title="Link to this heading"></a></h2>
<p><strong>1. Set up analysis file</strong></p>
<p>First, initialize your sample data for analysis:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sample</span> <span class="o">=</span> <span class="n">methyl_data</span>
</pre></div>
</div>
<p><strong>2. Set Up Reference Data</strong></p>
<p>Ensure the reference data matches the sample array type:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reference</span> <span class="o">=</span> <span class="n">MethylData</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">REFERENCE_DIR</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Confirm array types match</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">sample</span><span class="o">.</span><span class="n">array_type</span> <span class="o">==</span> <span class="n">reference</span><span class="o">.</span><span class="n">array_type</span>
</pre></div>
</div>
<p>Alternatively if the reference directory contains IDAT files of multiple
array types you can use ReferenceMethylData to load all files into memory.
The CNV class will then extract the files of the needed array.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">reference_all</span> <span class="o">=</span> <span class="n">ReferenceMethylData</span><span class="p">(</span><span class="n">REFERENCE_DIR</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>3. Initialize CNV Analysis</strong></p>
<p>Create an instance of the CNV class for analysis, and fit data (linear
regression model):</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span> <span class="o">=</span> <span class="n">CNV</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">reference</span><span class="p">)</span>
</pre></div>
</div>
<p>Alternative with ReferenceMethylData</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span> <span class="o">=</span> <span class="n">CNV</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">reference_all</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>4. Calculate CNV for Bins and Genes</strong></p>
<p>Compute CNV values for genomic bins and gene regions:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span><span class="o">.</span><span class="n">set_bins</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span><span class="o">.</span><span class="n">set_detail</span><span class="p">()</span>
</pre></div>
</div>
<p><strong>5. Calculate CNV Segments</strong></p>
<p>Use the binary circular segmentation algorithm for genome segmentation:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span><span class="o">.</span><span class="n">set_segments</span><span class="p">()</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For this step, additional packages must be installed (see
installation).</p>
</div>
<p><strong>6. Streamlined Analysis</strong></p>
<p>Alternatively, perform all CNV computations in a single call:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span> <span class="o">=</span> <span class="n">CNV</span><span class="o">.</span><span class="n">set_all</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">reference</span><span class="p">)</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c1"># or</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span> <span class="o">=</span> <span class="n">CNV</span><span class="o">.</span><span class="n">set_all</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">reference_all</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>7. Visualize CNV Data</strong></p>
<p>Display an interactive plot using Plotly, where genes can be highlighted:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cnv</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api.html" class="btn btn-neutral float-right" title="API Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Jon Brugger.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>